### DISCUSSION TITLE
Performance guidance

### DISCUSSION THREAD

**Discussion by jps1** [pyodide/pyodide#1406] [General]
I'm working on a project using pyodide and have a few different performance-related questions:

- Do you have any metrics / general feel for performance of doing a micropip install vs. building a custom pyodide image, with the .js + .data for the package pre-built?  The .whl is presumably smaller/faster to download, but then I'm not sure how the time for doing the "install" compares to the time extracting a "pre-compiled" .data file to the emscripten virtual filesystem (note: my understanding of what's going on behind-the-scenes here is rudimentary at best... I've gleaned what I can from [this documentation](https://pyodide.org/en/latest/development/new-packages.html#structure-of-a-pyodide-package)).  A [response to another question](https://github.com/pyodide/pyodide/discussions/1397#discussioncomment-553730) mentioned the possibility of pure python packages being installed exclusively with micropip, but I'm wondering if that would have perf implications.
- Have you done any profiling or other testing to understand at a high level the distribution of time being spent on various parts of pyodide workflows?  Is the lion's share of the time spent loading the python runtime?  Writing .data's into the emscripten virtual filesystem?  Just general, across-the-board slow-down due to "extra layers of abstraction"?  Is this the sort of thing where this kind of workflow is still fairly new, and there's likely to be a fair amount of low-hanging fruit that might yield significant speed-ups (like with the newer version of emscripten in 0.17.x), or are there fundamental limitations to running in the browser / WASM sandbox, etc, that make it this unlikely?
- Is there any way to "pre-build" an image with packages already in the main pyodide.asm.data?  Would there be any performance benefit to that? (kind've related to the previous question about what takes the time here).
- Is the focus of pyoyide as a project more towards larger, "notebook"-type workflows?  I've been playing around with using it for custom results-visualization within in an existing site, but loading a separate iframe with a new pyodide instance every time is fairly painful when it takes ~12-15 seconds to draw a single plot.  If this were on the order of 2-3 seconds instead (still many times slower than native python time to draw an equivalent matplotlib plot, so that doesn't _seem_ like an unreasonable goal), that's probably fast enough, but as is, I'm wondering whether I need to be loading the runtime once and then hiding/reusing the same frame.  I'd prefer to avoid this if I can get away with it, but I'm not sure I can.
- Do you have any experience as to whether choice of browser makes a significant performance difference when using pyodide?  I mostly stick to Firefox, but people obviously have their own preferences, so it would be good to know whether that's something I should be aware of.

Thanks!

**Comment by hoodmane** [comment]
A lot of interesting questions here. I don't know the answer to most of them.

> Is this the sort of thing where this kind of workflow is still fairly new, and there's likely to be a fair amount of low-hanging fruit that might yield significant speed-ups (like with the newer version of emscripten in 0.17.x), or are there fundamental limitations to running in the browser / WASM sandbox, etc, that make it this unlikely?

I'm not that knowledgeable about these issues, but I think there is likely a lot of low-hanging fruit that could yield speedups. There are main fundamental limitation of the browser sandbox is that we can't really have persistent installed software.

> but loading a separate iframe with a new pyodide instance every time is fairly painful when it takes ~12-15 seconds to draw a single plot. If this were on the order of 2-3 seconds instead (still many times slower than native python time to draw an equivalent matplotlib plot, so that doesn't seem like an unreasonable goal), that's probably fast enough, but as is, I'm wondering whether I need to be loading the runtime once and then hiding/reusing the same frame. I'd prefer to avoid this if I can get away with it, but I'm not sure I can.

What you should remember is that when Pyodide loads it's approximately similar to installing a copy of Python (and also matplotlib and whatever packages you need). I think that's the main limitation is that you have to install a new copy every time. I think the memory usage alone from loading lots of short-lived copies of Pyodide into the browser might be prohibitive.

I would suggest running Pyodide in a webworker and rendering the plot into an ArrayBuffer then transferring the ArrayBuffer to the main browser thread using the [postMessage transfer argument](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#syntax).

**Reply by jps1** [reply]
> What you should remember is that when Pyodide loads it's approximately similar to installing a copy of Python (and also matplotlib and whatever packages you need). I think that's the main limitation is that you have to install a new copy every time. I think the memory usage alone from loading lots of short-lived copies of Pyodide into the browser might be prohibitive.
> 
> I would suggest running Pyodide in a webworker and rendering the plot into an ArrayBuffer then transferring the ArrayBuffer to the main browser thread using the [postMessage transfer argument](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#syntax).

That's a really interesting point about being similar to installing a copy of python + libraries, something I hadn't really thought about.  Thanks for the suggestion about running in a webworker; I'll look into that.

**Reply by jps1** [reply]
I haven't had much time yet to investigate using a webworker, but this did get me thinking... is there a reason pyodide couldn't use emscripten's [idbfs filesystem API](https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-idbfs) or something like that to store the various library packages so they could persist across sessions, etc?  I'm pretty unfamiliar with this area, so maybe there's a really obvious reason this won't work, but if a significant chunk of time is spent just essentially creating a new Python environment every time, perhaps there's some way to avoid that.

**Reply by jps1** [reply]
To follow up on my own question here...
It appears that [IndexedDB no longer supports WASM modules](https://github.com/pyodide/pyodide/issues/806).  Also, there was some discussion of persistence of modules [here](https://github.com/pyodide/pyodide/issues/242), though it sounds like reproducibility / starting from a "clean slate" is weighted more highly than performance.  This is rather disappointing, since there's no point in being able to reproduce anything if the scenario can't be made fast enough in the first place (or it's sufficiently complicated to do so that the developer gives up).

**Reply by rth** [reply]
It depends if you are interested in stdlib or numeric packages. For the latter #617 should be a low hanging fruit.

**Reply by jps1** [reply]
Just to make sure I'm understanding... it sounds like the exclusion of the unit tests for third party modules would likely achieve somewhere on the order of 20% reduction in size, which would presumably result in a similar reduction in startup time (assuming the lions' share of startup time is spent loading files into the virtual fs, which seems likely, but isn't a given since there aren't specific benchmarks); is that correct?

I would still think investigation of persistence would be important.  A 20% reduction would be great (and would help for the "first-load" experience), but even that isn't enough to bring startup time down to a level where one doesn't really have to consider it that much.  Plus that would be something that would help across-the-board, rather than having to potentially continuously address new things that appear.

**Reply by jps1** [reply]
So I started working my way through getting matplotlib drawing something in a web-worker and hit issue #442.  The workaround there unblocked me, but as I'm working through this more, I'm increasingly questioning this approach.  I need to be able to interact with the plots (hovering over lines, zooming, using mouse motion events to add annotation, etc), and while I imagine it's _possible_ to do this in a web-worker, I'm starting to think all the plumbing stuff back and forth is going to be quite complicated.

I've hit a few bumps getting stuff working in the "non-web-worker" case, but all told it was relatively pain-free for the functionality I needed.  However, given the slowness of loading some of the larger python libraries every time, that approach doesn't seem viable either.

I'm thinking maybe I need to abandon matplotlib completely, use pyodide in web-worker just for some of the basic data-manipulation, and then hand off to d3 or some other javascript library for doing the visualization.  This is a shame because I and most of my users are already familiar with matplotlib (and it was super nice and easy to be able to play around with a normal, client-side python version of my scripts to make changes, then run it in pyodide + browser!), but it seems like for anything past plotting static figures, this may not be practical.

Am I being too pessimistic here, and it'll actually be easier than I thought?  Are there some more complex examples that would alleviate my concerns that I'm fighting upstream?  Or are you familiar with pyodide users that are doing this sort of thing?

**Reply by hoodmane** [reply]
> I need to be able to interact with the plots (hovering over lines, zooming, using mouse motion events to add annotation, etc), and while I imagine it's possible to do this in a web-worker, I'm starting to think all the plumbing stuff back and forth is going to be quite complicated.

Are you using comlink? 
https://github.com/GoogleChromeLabs/comlink#readme
With comlink and a little care, I think it would be relatively reasonable to adapt `wasm_backend.py` to `wasm_webworker_backend.py`. You would need to rewrite the whole thing and split it into a main thread part in Javascript and a pyodide part and that split would take a fair amount of extra care but my guess is that it would be only a modest increase in complexity. In particular, you wouldn't have to write any message-passing code directly.

**Reply by hoodmane** [reply]
For Pyodide version 18, one of my goals is to add much better support for webworkers. My plan is to base the entire thing on Comlink. matplotlib in a webworker is one feature that I will definitely be looking in to.

**Reply by hoodmane** [reply]
> Am I being too pessimistic here, and it'll actually be easier than I thought?

Yes, but only because Comlink simplifies things incredibly.

**Reply by jps1** [reply]
I was not familiar with Comlink.  Looks very interesting!  Thanks

**Reply by hoodmane** [reply]
I have code that uses Comlink here which might help to get you started9:
https://github.com/hoodmane/worker-pyodide-console
That's a test that sets up blocking IO in a worker.
Can see the result here, though it's just another repl.
https://spectralsequences.com/worker-pyodide-console/

**Comment by hoodmane** [comment]
> Do you have any experience as to whether choice of browser makes a significant performance difference when using pyodide?

Taking geometric means across our recent benchmarks shows that Firefox is 10-15% faster than Chrome.

**Comment by hoodmane** [comment]
> Is there any way to "pre-build" an image with packages already in the main pyodide.asm.data?

Yes: the pyodide package is preloaded with:
https://github.com/pyodide/pyodide/blob/5b405767f957cadd935ac2197090bb784b2a2d27/Makefile#L52
You can add more lines like that to the makefile to preload more files.

**Comment by rth** [comment]
To add to @hoodmane 's answer https://github.com/pyodide/pyodide/discussions/1406#discussioncomment-555447

Generally you might find https://emscripten.org/docs/compiling/Deploying-Pages.html useful  also see related discussions in https://github.com/pyodide/pyodide/issues/1365 https://github.com/pyodide/pyodide/issues/347

> Do you have any metrics / general feel for performance of doing a micropip install vs. building a custom pyodide image, with the .js + .data for the package pre-built?

No detailed benchmarks have been done for this as far as I know. I would expect them to be roughly equivalent both in size and unpacking time, but benchmarks are needed to be sure. Fundamentally they are identical .py files compressed (as zip for wheels) or (LZ4 for .data), that then gets additionally compressed by the CDN as gzip or brotli and that need to be unpacked into the virtual file system.

 > Is this the sort of thing where this kind of workflow is still fairly new, and there's likely to be a fair amount of low-hanging fruit that might yield significant speed-ups

Yes, I would say it's likely. And we need better profiling of different steps to start. Logging and timing file system access could also be useful https://docs.huihoo.com/emscripten/emscripten-site/docs/porting/files/packaging_files.html#monitoring-file-usage

> Is there any way to "pre-build" an image with packages already in the main pyodide.asm.data?

https://emscripten.org/docs/compiling/Deploying-Pages.html suggests rather going in the other direction for performance, splitting data files and loading them in parallel.
